services/datamanager/builtin/file_deletion_test.go

```diff
@@ -0,0 +1,320 @@
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			tempCaptureDir := t.TempDir()
+			logger := logging.NewTestLogger(t)
+			mockClient := mockDataSyncServiceClient{
+				succesfulDCRequests: make(chan *v1.DataCaptureUploadRequest, 100),
+				failedDCRequests:    make(chan *v1.DataCaptureUploadRequest, 100),
+				fail:                &atomic.Bool{},
+			}
+
+			var syncer datasync.Manager
+			if tc.syncEnabled {
+				s, err := datasync.NewManager("rick astley", mockClient, logger, tempCaptureDir)
+				test.That(t, err, test.ShouldBeNil)
+				syncer = s
+				defer syncer.Close()
+			}
+
+			filepaths := writeFiles(t, tempCaptureDir, tc.fileList)
+			for _, file := range tc.syncerInProgressFiles {
+				syncer.MarkInProgress(filepaths[file])
+			}
+
+			ctx, cancelFunc := context.WithCancel(context.Background())
+			defer cancelFunc()
+			if tc.shouldCancelContext {
+				cancelFunc()
+			}
+			deletedFileCount, err := deleteFiles(ctx, syncer, tempCaptureDir, logger)
+			if tc.shouldCancelContext {
+				test.That(t, err, test.ShouldBeError, context.Canceled)
+			} else {
+				test.That(t, err, test.ShouldBeNil)
+				test.That(t, deletedFileCount, test.ShouldEqual, tc.expectedDeletedCount)
+			}
+		})
+	}
+}
+
+func writeFiles(t *testing.T, dir string, filenames []string) map[string]string {
+	t.Helper()
+	fileContents := []byte("never gonna let you down")
+	filePaths := map[string]string{}
+	for _, filename := range filenames {
+		filePath := fmt.Sprintf("%s/%s", dir, filename)
+		err := os.WriteFile(filePath, fileContents, 0o755)
+		test.That(t, err, test.ShouldBeNil)
+		filePaths[filename] = filePath
+	}
+	return filePaths
+}
+
+func TestFilePolling(t *testing.T) {
+	mockClock := clk.NewMock()
+	// Make mockClock the package level clock used by the dmsvc so that we can simulate time's passage
+	clock = mockClock
+	deletionTicker = mockClock
+	filesystemPollInterval = time.Millisecond * 20
+
+	tempDir := t.TempDir()
+	fsThresholdToTriggerDeletion = math.SmallestNonzeroFloat64
+	captureDirToFSUsageRatio = math.SmallestNonzeroFloat64
+
+	// Set up data manager.
+	dmsvc, mockClient := newDMSvc(t, tempDir)
+
+	mockClock.Add(captureInterval)
+	flusher, ok := dmsvc.(*builtIn)
+	test.That(t, ok, test.ShouldBeTrue)
+	// flush and close collectors to ensure we have exactly 4 files
+	flusher.flushCollectors()
+	flusher.closeCollectors()
+
+	waitForCaptureFilesToExceedNFiles(tempDir, 4)
+
+	files := getAllFileInfos(tempDir)
+	test.That(t, len(files), test.ShouldEqual, 4)
+	// since we've written 4 files and hit the threshold, we expect
+	// the first to be deleted
+	expectedDeletedFile := files[0]
+
+	mockClock.Add(filesystemPollInterval)
+	newFiles := getAllFileInfos(tempDir)
+	test.That(t, len(newFiles), test.ShouldEqual, 3)
+	test.That(t, newFiles, test.ShouldNotContain, expectedDeletedFile)
+	// advance by time remaining to hit 1 sync interval
+	mockClock.Add(syncInterval - (filesystemPollInterval + captureInterval))
```

Comment:
To make it clear what's happening, it might be worth noting that the capture interval is 10ms and sync interval is 50ms from the other builtin test files. So we run forward 10ms to capture 4 files then close the collectors, run forward 20ms to delete any files, then run forward 20ms to sync the remaining 3 files.